aiida/tools/visualization/graph.py

```diff
@@ -564,55 +661,88 @@ def recurse_descendants(
-                )
-                if outgoing_nodes and print_func:
-                    print_func('  {} -> {}'.format(node.pk, [on.pk for on in outgoing_nodes]))
-                new_nodes.extend(outgoing_nodes)
-
-                if include_process_inputs and isinstance(node, ProcessNode):
-                    self.add_incoming(node, link_types=link_types, annotate_links=annotate_links)
+        # pylint: disable=too-many-arguments,too-many-locals
+        import warnings
+        from aiida.common.warnings import AiidaDeprecationWarning
+        if print_func:
+            warnings.warn( # pylint: disable=no-member
+                '`print_func` is deprecated because graph traversal has been refactored', AiidaDeprecationWarning
+            )
 
-            # ensure the same path isn't traversed multiple times
-            leaf_nodes = []
-            for new_node in new_nodes:
-                if new_node.pk in traversed_pks:
-                    continue
-                leaf_nodes.append(new_node)
-                traversed_pks.append(new_node.pk)
+        # Get graph traversal rules where the given link types and direction are all set to True, and all others are
+        # set to False
+        origin_pk = origin if isinstance(origin, int) else origin.pk
+        graph_traversal_rules = self._link_types_to_traversal_rules(link_types=link_types, direction='forward')
+
+        # Use the graph traverser to get all nodes and links for the vizualization
+        params = {'max_iterations': depth, 'get_links': True}
+        parsed_rules = parse_traversal_rules(**graph_traversal_rules)
+        params['links_forward'] = parsed_rules['forward']
+        params['links_backward'] = parsed_rules['backward']
+        traversed_graph = traverse_graph((origin_pk,), **params)
+
+        # Traverse backward along input_work and input_calc links from all nodes traversed in the previous step
+        # and join the result with the original traversed graph. This includes calculation inputs in the Graph
+        if include_process_inputs:
+            traversed_outputs = traverse_graph(
+                traversed_graph['nodes'],
+                max_iterations=1,
+                get_links=True,
+                links_backward=[LinkType.INPUT_WORK, LinkType.INPUT_CALC]
+            )
+            traversed_graph['nodes'] = traversed_graph['nodes'].union(traversed_outputs['nodes'])
+            traversed_graph['links'] = traversed_graph['links'].union(traversed_outputs['links'])
+
+        # Do one central query for all nodes in the Graph and generate a {id: Node} dictionary
+        traversed_nodes = orm.QueryBuilder().append(
+            orm.Node,
+            filters={'id': {
+                'in': traversed_graph['nodes']
+            }},
+            project=['id', '*'],
+            tag='node',
+        )
+        traversed_nodes = {query_result[0]: query_result[1] for query_result in traversed_nodes.all()}
```

Comment:
would
```
traversed_nodes = {pk: load_node(pk) for pk in traversed_graph['nodes']}
```
be slower? If not, it would be a lot clearer, no?

Reply:
Actually, wouldn't it be...faster? I mean, when projecting the node you are loading it anyways, which is the slow part, correct?